apps/src/StudioApp.js

```diff
@@ -843,17 +868,96 @@ StudioApp.prototype.createModalDialog = function(options) {
   return this.feedback_.createModalDialog(options);
 };
 
+StudioApp.prototype.recordUserViewedHint_ = function (hint, config) {
+  hint.alreadySeen = true;
+
+  authoredHintUtils.recordUnfinishedHint({
+    // level info
+    script_id: config.scriptId,
+    level_id: config.level.scriptLevelId,
+
+    // hint info
+    hint_id: hint.hintId,
+    hint_class: hint.hintClass,
+    hint_type: hint.hintType,
+  });
+};
+
+StudioApp.prototype.onReportComplete = function (response) {
+  authoredHintUtils.finishHints({
+    time: ((new Date().getTime()) - this.initTime),
+    attempt: this.attempts,
+    test_result: this.lastTestResult,
+    activity_id: response && response.activity_id,
+    level_source_id: response && response.level_source_id,
+  });
+};
+
+StudioApp.prototype.showHint_ = function (hint, config) {
+
```

Comment:
Worth some param comments describing the shape of hint/config? Also, do we need to pass the entire config through. If we're only using a couple of fields from it, it might be better to pass just those fields so that it's clearer which parts of config we depend on